webrtc.html

```diff
@@ -4790,23 +4770,43 @@ <h2>Methods</h2>
               <p>The resulting certificate MUST NOT include information that
               can be linked to a user or <a>user agent</a>. Randomized values
               for distinguished name and serial number SHOULD be used.</p>
-              <p>A <a>user agent</a> MUST reject a call to
-              <code>generateCertificate()</code> with a
-              <code>DOMException</code> of type <code>NotSupportedError</code>
-              if the <var>keygenAlgorithm</var> parameter identifies an
-              algorithm that the <a>user agent</a> cannot or will not use to
-              generate a certificate for <code>RTCPeerConnection</code>.</p>
-              <p>The following values MUST be supported by a <a>user agent</a>:
-              <code>{ name: "<a data-cite=
-              "!WebCryptoAPI#rsassa-pkcs1">RSASSA-PKCS1-v1_5</a>",
-              modulusLength: 2048, publicExponent: new Uint8Array([1, 0, 1]),
-              hash: "SHA-256" }</code>, and <code>{ name: "<a data-cite=
-              "!WebCryptoAPI#ecdsa">ECDSA</a>",
-              namedCurve: "<a data-cite=
-              "!WebCryptoAPI#dfn-NamedCurve">P-256</a>"
-              }</code>.</p>
-              <p class="note">It is expected that a <a>user agent</a> will have
-              a small or even fixed set of values that it will accept.</p>
+
+              <p>When the method is called, the user agent MUST run the following steps:</p>
+              <ol>
+                <li>Let <var>keygenAlgorithm</var> be the first argument to <code>generateCertificate</code>.</li>
+                <li>Let <var>certificateExpiration</var> be the result of converting <var>keygenAlgorithm</var> to an <code>RTCCertificateExpiration</code> value.</li>
```

Comment:
This makes no sense. You can't convert keygenAlgorithm to an expiration value. Wrong variable?

Reply:
This tries to implement what is being defined in https://w3c.github.io/webrtc-pc/#rtccertificateexpiration-dictionary.

I agree that this is not great. I thought about changing the IDL to use any instead of AlgorithmIdentifier there but kept it this way for now.

The idea behind the current description is that "keygenAlgorithm" being the first argument to generateCertificate, it is not yet typed as a AlgorithmIdentifier.

Reply:
I changed the WebIDL to have generateCertificate take 'any' instead and do the conversion twice.

Reply:
@alvestrand @youennf  I think we should stick with [AlgorithmIdentifier](https://w3c.github.io/webcrypto/#dfn-AlgorithmIdentifier) in the WebIDL. It is defined as:
```js
typedef (object or DOMString) AlgorithmIdentifier;
```
So when the old language says *"When generateCertificate is called with an `object` argument, the user agent attempts to convert the object into an RTCCertificateExpiration."*, it's not unclear how it can be an `object`—it can.

The unclear part is perhaps how to "convert" said `object` to said dictionary. Linking to https://heycam.github.io/webidl/#dfn-convert-ecmascript-to-idl-value should clarify this.

We always want to do as much in WebIDL vs. prose as possible, so we should avoid `any`.

Reply:
I meant to reply here, but clicked "Review", so see below instead.

Reply:
OK, I will revert the related changes back to the fist commit and will add an "if object" check to make it clear that conversion happens only in that case.

We can try to use 
https://heycam.github.io/webidl/#dfn-convert-ecmascript-to-idl-value but we would need to convert back the IDL value (which is an IDL object) in an ECMAScript value first (https://heycam.github.io/webidl/#dfn-convert-idl-to-ecmascript-value would do that I guess).
This is more or less matching what WebKit is currently doing so I am fine with that.

I wonder whether we can write a test for this. Is Chrome or Firefox supporting generateCertificate with a string value?

Reply:
> we would need to convert back the IDL value (which is an IDL object)

Technically. Hopefully [scant difference](https://heycam.github.io/webidl/#es-object) in practice.

I see [crypto](https://w3c.github.io/webcrypto/#dfn-normalize-an-algorithm) uses: *"Let initialAlg be the result of converting the ECMAScript object represented by alg to the IDL dictionary type Algorithm, as defined by [WebIDL]. "*

> Is Chrome or Firefox supporting generateCertificate with a string value?

Both pass the [wpt tests](https://wpt.fyi/results/webrtc/RTCPeerConnection-generateCertificate.html?label=master&product=chrome%5Bexperimental%5D&product=edge&product=firefox%5Bexperimental%5D&product=safari%5Bexperimental%5D&aligned) but I only see a test for `'invalid-algo'`. Unsure what plain strings are valid, but I managed to get to `NotSupportedError` in both with:
```js
await RTCPeerConnection.generateCertificate('P-256');
```
I guess some extra input validation tests may be in order.